lens/history.go

```diff
@@ -0,0 +1,279 @@
+		wasFound := false
+		for {
+			if !currentItem.next.HasValue() {
+				break
+			}
+
+			currentItem = currentItem.next.Value()
+			distanceTravelled++
+			if currentItem.targetVector != 0 {
+				distanceTravelled += currentItem.targetVector
+				wasFound = true
+				break
+			}
+			if currentItem.schemaVersionID == targetSchemaVersionID {
+				wasFound = true
+				break
+			}
+		}
+
+		if !wasFound {
+			// The target was not found going up the chain, try looking back.
+			// This is important for downgrading schema versions.
+			for {
+				if !currentItem.previous.HasValue() {
+					break
+				}
+
+				currentItem = currentItem.previous.Value()
+				distanceTravelled--
+				if currentItem.targetVector != 0 {
+					distanceTravelled += currentItem.targetVector
+					wasFound = true
+					break
+				}
+				if currentItem.schemaVersionID == targetSchemaVersionID {
+					wasFound = true
+					break
+				}
+			}
+		}
+
+		if !wasFound {
+			// This may happen if users define schema migrations to unknown schema versions
+			// with no migration path to known schema versions, esentially creating orphan
+			// migrations. These may become linked later and should remain persisted in the
+			// database, but we can drop them from the history here/now.
+			orphanSchemaVersions[schemaVersion] = struct{}{}
+			continue
+		}
+
+		item.targetVector = distanceTravelled
+	}
+
+	for schemaVersion := range orphanSchemaVersions {
+		delete(result, schemaVersion)
+	}
+
+	return result, nil
+}
+
+type schemaHistoryPairing struct {
+	schemaVersionID     string
+	nextSchemaVersionID string
+}
+
+// getSchemaHistory returns the history of the schema of the given id as linked list
+// with each item mapped by schema version id.
+//
+// This includes any history items that are only known via registered
+// schema migrations.
+func getSchemaHistory(
+	ctx context.Context,
+	txn datastore.Txn,
+	lensConfigs []client.LensConfig,
+	schemaID string,
+) (map[schemaVersionID]*schemaHistoryLink, error) {
+	pairings := map[string]*schemaHistoryPairing{}
+
+	for _, config := range lensConfigs {
+		pairings[config.SourceSchemaVersionID] = &schemaHistoryPairing{
+			schemaVersionID:     config.SourceSchemaVersionID,
+			nextSchemaVersionID: config.DestinationSchemaVersionID,
+		}
+
+		if _, ok := pairings[config.DestinationSchemaVersionID]; !ok {
+			pairings[config.DestinationSchemaVersionID] = &schemaHistoryPairing{
+				schemaVersionID: config.DestinationSchemaVersionID,
+			}
+		}
+	}
+
+	prefix := core.NewSchemaHistoryKey(schemaID, "")
+	q, err := txn.Systemstore().Query(ctx, query.Query{
+		Prefix: prefix.ToString(),
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	for res := range q.Next() {
+		// check for Done on context first
+		select {
+		case <-ctx.Done():
+			// we've been cancelled! ;)
+			return nil, q.Close()
+		default:
+			// noop, just continue on the with the for loop
+		}
+
+		if res.Error != nil {
+			err = q.Close()
+			if err != nil {
+				return nil, err
```

Comment:
todo: why are you prioritising `q.Close()` error over `res.Error` which happened first?

We should return the first error that happened.

Also applied for the code below

Reply:
> We should return the first error that happened.

Why do you think that the first error is the most important one? `Close` failing could be far more important to both us and the user.

Reply:
you are right, we don't know which one of the might be more important. Error on close can expose some severe issues while the first error (as a first "witness") can point out the origin. Or the other way around.
In this case the best would be the wrap one error into another, so we don't loose any context.

Reply:
I disagree here, the original error feels more important, but that isn't to say we should be ignoring either. Classic case for why the upcoming `errors.Join` is so helpful.

A short term would be to wrap the original error with a stringified close error.

Basically:
```go
if res.Error != nil {
  err = q.Close()
  if err != nil {
    return nil, errors.Wrap(err.Error(), res.Error)
  }
}
```

this is a poor mans `errors.Join` because we lose the error type information from the closing error. This can technically be implemented in the `errors` package to make it look/feel a little cleaner.

Reply:
> I disagree here

With what? :grin: 

Returning both errors is of course better, this is a fairly common code-crime in our codebase which we should correct at some point.

- [ ] Return both

Reply:
Disagree that the close error is more important.

Reply:
> Disagree that the close error is more important.

I didnt say that it was, I only pointed out that it can be, and picking one over the other was fairly arbitrary.

Reply:
> Why do you think that the first error is the most important one? Close failing could be far more important to both us and the user.

Apologies, interpreted this as your preference in choosing close over the root error